### PR TITLE
Feature: Rep contract

### LINKED ISSUE #81: feat-reputration-contract
## User Story
As a DAO member, I want to be able to vote for proposals based on my reputation.

## Status Quo
Reputation is used in the legacy DAOStack DAO but is not integrated in the current architecture.  We have to create a new non-transferable ERC20 token that will copy reputation from the legacy DAO rep.

## Acceptance Criteria
Non-transferable ERC20 contract, that has the functionality of batched REP minting

### PR DESCRIPTION
**Pull request by fabianschu**
❗ ❗ ❗ 
**The Reputation contract should not be deployed before we have the correct initial REP distribution. (#88)** 
❗ ❗ ❗ 

[These](https://github.com/PrimeDAO/contracts-v2/blob/81b-rep-contract/inputs/initialRepBalances.json) are only placeholder values. The real initial distributions still need to be added (#88). The two addresses in the example file `inputs/initialRepBalances.json` are account 2 & 3 of our shared development mnemonic. Accordingly these contracts hold 1.5 and 1 REP on rinkeby. Account 1 is the owner of the contract.

The Reputation contract is deployed to Rinkeby, verified on Etherscan and can hence be interacted with via Etherscan. (If you connect through Account 1 of our dev mnemonic to Etherscan, you can mint/burn tokens for further testing). Here is the [link](https://rinkeby.etherscan.io/address/0x5Da95f305858C3517913B5938Ae0fC49BBb18453#code).

A Snapshot location has been configured to use the deployed Rinkeby REP tokens for voting. Please find a test proposal [here](https://snapshot.org/#/fuschu.eth/proposal/QmXAGrUHdxcRwmSNhWajAVp2646uoSUhowL3nvLyeh1EEg).

closes #81

### CODE REVIEW

**Comment by fabianschu** [comment]
This is a simulation on the estimated gas necessary to deploy the contract, depending on the length of arrays provided to the constructor. For reference  `blockGasLimit = 12450000`.
```
┌─────────┬─────────────┬──────────────┬──────────────────────────────┬──────────────────────────────┬───────────────────────────────┬───────────────────────────────┐
│ (index) │ arrayLength │ estimatedGas │ gasCostsInEth_gasPrice10Gwei │ gasCostsInEth_gasPrice50Gwei │ gasCostsInEth_gasPrice100Gwei │ gasCostsInEth_gasPrice300Gwei │
├─────────┼─────────────┼──────────────┼──────────────────────────────┼──────────────────────────────┼───────────────────────────────┼───────────────────────────────┤
│    0    │      5      │   2232541    │         '0.02232541'         │         '0.11162705'         │          '0.2232541'          │          '0.6697623'          │
│    1    │     10      │   2255433    │         '0.02255433'         │         '0.11277165'         │          '0.2255433'          │          '0.6766299'          │
│    2    │     15      │   2278313    │         '0.02278313'         │         '0.11391565'         │          '0.2278313'          │          '0.6834939'          │
│    3    │     20      │   2301193    │         '0.02301193'         │         '0.11505965'         │          '0.2301193'          │          '0.6903579'          │
│    4    │     25      │   2324073    │         '0.02324073'         │         '0.11620365'         │          '0.2324073'          │          '0.6972219'          │
│    5    │     30      │   2346953    │         '0.02346953'         │         '0.11734765'         │          '0.2346953'          │          '0.7040859'          │
│    6    │     35      │   2369833    │         '0.02369833'         │         '0.11849165'         │          '0.2369833'          │          '0.7109499'          │
│    7    │     40      │   2392713    │         '0.02392713'         │         '0.11963565'         │          '0.2392713'          │          '0.7178139'          │
│    8    │     45      │   2415581    │         '0.02415581'         │         '0.12077905'         │          '0.2415581'          │          '0.7246743'          │
│    9    │     50      │   2438473    │         '0.02438473'         │         '0.12192365'         │          '0.2438473'          │          '0.7315419'          │
│   10    │     55      │   2461353    │         '0.02461353'         │         '0.12306765'         │          '0.2461353'          │          '0.7384059'          │
│   11    │     60      │   2484233    │         '0.02484233'         │         '0.12421165'         │          '0.2484233'          │          '0.7452699'          │
│   12    │     65      │   2507113    │         '0.02507113'         │         '0.12535565'         │          '0.2507113'          │          '0.7521339'          │
│   13    │     70      │   2530017    │         '0.02530017'         │         '0.12650085'         │          '0.2530017'          │          '0.7590051'          │
│   14    │     75      │   2552981    │         '0.02552981'         │         '0.12764905'         │          '0.2552981'          │          '0.7658943'          │
│   15    │     80      │   2575947    │         '0.02575947'         │         '0.12879735'         │          '0.2575947'          │          '0.7727841'          │
│   16    │     85      │   2598902    │         '0.02598902'         │         '0.1299451'          │          '0.2598902'          │          '0.7796706'          │
│   17    │     90      │   2621882    │         '0.02621882'         │         '0.1310941'          │          '0.2621882'          │          '0.7865646'          │
│   18    │     95      │   2644852    │         '0.02644852'         │         '0.1322426'          │          '0.2644852'          │          '0.7934556'          │
│   19    │     100     │   2667824    │         '0.02667824'         │         '0.1333912'          │          '0.2667824'          │          '0.8003472'          │
│   20    │     105     │   2690797    │         '0.02690797'         │         '0.13453985'         │          '0.2690797'          │          '0.8072391'          │
│   21    │     110     │   2713772    │         '0.02713772'         │         '0.1356886'          │          '0.2713772'          │          '0.8141316'          │
│   22    │     115     │   2736748    │         '0.02736748'         │         '0.1368374'          │          '0.2736748'          │          '0.8210244'          │
│   23    │     120     │   2759726    │         '0.02759726'         │         '0.1379863'          │          '0.2759726'          │          '0.8279178'          │
│   24    │     125     │   2782694    │         '0.02782694'         │         '0.1391347'          │          '0.2782694'          │          '0.8348082'          │
│   25    │     130     │   2805687    │         '0.02805687'         │         '0.14028435'         │          '0.2805687'          │          '0.8417061'          │
│   26    │     135     │   2828669    │         '0.02828669'         │         '0.14143345'         │          '0.2828669'          │          '0.8486007'          │
│   27    │     140     │   2851653    │         '0.02851653'         │         '0.14258265'         │          '0.2851653'          │          '0.8554959'          │
│   28    │     145     │   2874639    │         '0.02874639'         │         '0.14373195'         │          '0.2874639'          │          '0.8623917'          │
│   29    │     150     │   2897626    │         '0.02897626'         │         '0.1448813'          │          '0.2897626'          │          '0.8692878'          │
│   30    │     155     │   2920615    │         '0.02920615'         │         '0.14603075'         │          '0.2920615'          │          '0.8761845'          │
│   31    │     160     │   2943605    │         '0.02943605'         │         '0.14718025'         │          '0.2943605'          │          '0.8830815'          │
│   32    │     165     │   2966585    │         '0.02966585'         │         '0.14832925'         │          '0.2966585'          │          '0.8899755'          │
│   33    │     170     │   2989591    │         '0.02989591'         │         '0.14947955'         │          '0.2989591'          │          '0.8968773'          │
│   34    │     175     │   3012586    │         '0.03012586'         │         '0.1506293'          │          '0.3012586'          │          '0.9037758'          │
│   35    │     180     │   3035583    │         '0.03035583'         │         '0.15177915'         │          '0.3035583'          │          '0.9106749'          │
│   36    │     185     │   3058581    │         '0.03058581'         │         '0.15292905'         │          '0.3058581'          │          '0.9175743'          │
│   37    │     190     │   3081581    │         '0.03081581'         │         '0.15407905'         │          '0.3081581'          │          '0.9244743'          │
│   38    │     195     │   3104582    │         '0.03104582'         │         '0.1552291'          │          '0.3104582'          │          '0.9313746'          │
│   39    │     200     │   3127585    │         '0.03127585'         │         '0.15637925'         │          '0.3127585'          │          '0.9382755'          │
│   40    │     205     │   3150577    │         '0.03150577'         │         '0.15752885'         │          '0.3150577'          │          '0.9451731'          │
│   41    │     210     │   3173595    │         '0.03173595'         │         '0.15867975'         │          '0.3173595'          │          '0.9520785'          │
│   42    │     215     │   3196603    │         '0.03196603'         │         '0.15983015'         │          '0.3196603'          │          '0.9589809'          │
│   43    │     220     │   3219612    │         '0.03219612'         │         '0.1609806'          │          '0.3219612'          │          '0.9658836'          │
│   44    │     225     │   3242623    │         '0.03242623'         │         '0.16213115'         │          '0.3242623'          │          '0.9727869'          │
│   45    │     230     │   3265635    │         '0.03265635'         │         '0.16328175'         │          '0.3265635'          │          '0.9796905'          │
│   46    │     235     │   3288649    │         '0.03288649'         │         '0.16443245'         │          '0.3288649'          │          '0.9865947'          │
│   47    │     240     │   3311664    │         '0.03311664'         │         '0.1655832'          │          '0.3311664'          │          '0.9934992'          │
│   48    │     245     │   3334669    │         '0.03334669'         │         '0.16673345'         │          '0.3334669'          │          '1.0004007'          │
│   49    │     250     │   3357700    │          '0.033577'          │          '0.167885'          │           '0.33577'           │           '1.00731'           │
│   50    │     255     │   3380720    │         '0.0338072'          │          '0.169036'          │          '0.338072'           │          '1.014216'           │
│   51    │     260     │   3403765    │         '0.03403765'         │         '0.17018825'         │          '0.3403765'          │          '1.0211295'          │
│   52    │     265     │   3426789    │         '0.03426789'         │         '0.17133945'         │          '0.3426789'          │          '1.0280367'          │
│   53    │     270     │   3449813    │         '0.03449813'         │         '0.17249065'         │          '0.3449813'          │          '1.0349439'          │
│   54    │     275     │   3472840    │         '0.0347284'          │          '0.173642'          │          '0.347284'           │          '1.041852'           │
│   55    │     280     │   3495868    │         '0.03495868'         │         '0.1747934'          │          '0.3495868'          │          '1.0487604'          │
│   56    │     285     │   3518885    │         '0.03518885'         │         '0.17594425'         │          '0.3518885'          │          '1.0556655'          │
│   57    │     290     │   3541928    │         '0.03541928'         │         '0.1770964'          │          '0.3541928'          │          '1.0625784'          │
│   58    │     295     │   3564961    │         '0.03564961'         │         '0.17824805'         │          '0.3564961'          │          '1.0694883'          │
│   59    │     300     │   3587995    │         '0.03587995'         │         '0.17939975'         │          '0.3587995'          │          '1.0763985'          │
│   60    │     305     │   3611030    │         '0.0361103'          │         '0.1805515'          │          '0.361103'           │          '1.083309'           │
│   61    │     310     │   3634068    │         '0.03634068'         │         '0.1817034'          │          '0.3634068'          │          '1.0902204'          │
│   62    │     315     │   3657107    │         '0.03657107'         │         '0.18285535'         │          '0.3657107'          │          '1.0971321'          │
│   63    │     320     │   3680147    │         '0.03680147'         │         '0.18400735'         │          '0.3680147'          │          '1.1040441'          │
│   64    │     325     │   3703177    │         '0.03703177'         │         '0.18515885'         │          '0.3703177'          │          '1.1109531'          │
│   65    │     330     │   3726232    │         '0.03726232'         │         '0.1863116'          │          '0.3726232'          │          '1.1178696'          │
│   66    │     335     │   3749277    │         '0.03749277'         │         '0.18746385'         │          '0.3749277'          │          '1.1247831'          │
│   67    │     340     │   3772324    │         '0.03772324'         │         '0.1886162'          │          '0.3772324'          │          '1.1316972'          │
│   68    │     345     │   3795372    │         '0.03795372'         │         '0.1897686'          │          '0.3795372'          │          '1.1386116'          │
│   69    │     350     │   3818422    │         '0.03818422'         │         '0.1909211'          │          '0.3818422'          │          '1.1455266'          │
│   70    │     355     │   3841473    │         '0.03841473'         │         '0.19207365'         │          '0.3841473'          │          '1.1524419'          │
│   71    │     360     │   3864526    │         '0.03864526'         │         '0.1932263'          │          '0.3864526'          │          '1.1593578'          │
│   72    │     365     │   3887569    │         '0.03887569'         │         '0.19437845'         │          '0.3887569'          │          '1.1662707'          │
│   73    │     370     │   3910637    │         '0.03910637'         │         '0.19553185'         │          '0.3910637'          │          '1.1731911'          │
│   74    │     375     │   3933694    │         '0.03933694'         │         '0.1966847'          │          '0.3933694'          │          '1.1801082'          │
│   75    │     380     │   3956754    │         '0.03956754'         │         '0.1978377'          │          '0.3956754'          │          '1.1870262'          │
│   76    │     385     │   3979814    │         '0.03979814'         │         '0.1989907'          │          '0.3979814'          │          '1.1939442'          │
│   77    │     390     │   4002876    │         '0.04002876'         │         '0.2001438'          │          '0.4002876'          │          '1.2008628'          │
│   78    │     395     │   4025940    │         '0.0402594'          │          '0.201297'          │          '0.402594'           │          '1.207782'           │
│   79    │     400     │   4049006    │         '0.04049006'         │         '0.2024503'          │          '0.4049006'          │          '1.2147018'          │
└─────────┴─────────────┴──────────────┴──────────────────────────────┴──────────────────────────────┴───────────────────────────────┴───────────────────────────────┘
```

**Review by kinspotting — APPROVED** [review]

File: contracts/Reputation.sol [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    function transfer(
+        address recipient,
+        uint256 amount
+    ) public override pure returns(bool) {
+        return false;
+    }
+
+    function transferFrom(
```
Comment (Zitzak) [inline contracts/Reputation.sol:52]: For my understanding. Did you override the transfer/transferFrom function so you effectively disabled the function from  the original ERC20 contract?

File: tasks/utils/reputation.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const { utils } = require("ethers");
+const initialRepBalances = require("../../inputs/initialRepBalances.json");
+
+const { parseEther } = utils;
+
+const getReputationParams = () => {
+  let repHolders = [];
+  let repAmounts = [];
+  Object.entries(initialRepBalances).forEach(([address, amount]) => {
```
Comment (Zitzak) [inline tasks/utils/reputation.js:9]: Does this sort of simulated a mapping? So you push the values up the stack and then return a key:value pair?

File: contracts/Reputation.sol (hunk 1 above)
Comment (fabianschu) [reply contracts/Reputation.sol:52]: @kinspotting did that with exactly this intention :)

File: tasks/utils/reputation.js (hunk 2 above)
Comment (fabianschu) [reply tasks/utils/reputation.js:9]: The contract expects as input parameters two equally sized arrays with contract addresses and balances. The function generates these arrays from a JSON object.

File: contracts/Reputation.sol (hunk 1 above)
Comment (Zitzak) [reply contracts/Reputation.sol:52]: Great then I understood what was happening. Thanks

File: tasks/utils/reputation.js (hunk 2 above)
Comment (Zitzak) [reply tasks/utils/reputation.js:9]: Ahh oke. Thanks

Comment (fabianschu) [reply tasks/utils/reputation.js:9]: And returns both arrays

**Pull request merged**